src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -3198,8 +3202,13 @@ class CyclesRenderer final : public IECoreScenePreview::Renderer
 		void render() override
 		{
 			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
-
 			{
+				// Make sure to apply any scene/session parameters first so we only hard-reset once
+				// before any rendering starts as we cannot set these after rendering starts.
+				if( m_renderState != RENDERSTATE_RENDERING )
```

Comment:
What is the intention of this check? That we will only ever call `updateSessionParams()` once? If so, how does that interact with the code in `progress()` that sets `m_renderState = RENDERSTATE_STOPPED;`. The comment there seems to imply that will happen whenever an interactive render converges and stops sampling.

Reply:
> Thanks Alex, it definitely seems like this is something we need to simplify. I don't really understand what we're doing here still though. Could you step back a bit and explain clearly what the problem is (what constraints Cycles is placing on us), and what our strategy is?
> 
> Implementationwise, I wonder if these are possible and would help with clarity?
> 

Yep, so the goals of this draft PR (keyword draft! :) ) was to tackle the windows deadlock from @ericmehl 's https://github.com/GafferHQ/gaffer/pull/5100, and the comments in there probably explain my beginning process the best (copy-pasted):

There was talks/suggestions before that we could defer Cycles' scene/session creation to the first call to `render()` so we don't have the reset at all, but the current code as it stands needs the first reset so that the desired scene/session parameters can be set before rendering starts. I've not started this yet and the annoying thing is any calls to `option()` needs to be caching all of these somewhere without scene/session existing yet, and maybe other things I'm not thinking of.

> * Could we construct the session only once, the first time `render()` is called, so everything is deferred till that moment, instead of trying to reset things that were done earlier?

I eventually want to move it to `render()` if possible but I still think it'll be a big headache before that. The biggest one is the need for `scene` to exist when converting objects and shaders eg. `scene->shader_manager` usage, and I know when we get VDB's to work again, `scene->image_manager` is needed as well. This PR just has better intentions where we only allow all session/scene parameter changes before rendering starts, and not the possibility of those changing while interactive rendering like it was intended in the past. The warning in `option()` was just a print-out and nothing more.

> * Could we query Cycles for what is changeable and what is not, so we're not duplicating information?
> 
https://github.com/blender/cycles/blob/master/src/session/session.h#L90 - this is the best I've found, the omission of some parameters in the modified() call.

> Some test coverage for all this would be pretty useful too...

After draft status no worries

Reply:
> What is the intention of this check? That we will only ever call `updateSessionParams()` once? If so, how does that interact with the code in `progress()` that sets `m_renderState = RENDERSTATE_STOPPED;`. The comment there seems to imply that will happen whenever an interactive render converges and stops sampling.

Alright going over the enum states I've just now made a `bool m_rendering`, the states were needed pre-CyclesX when a finished interactive render would cease to update, but that's not needed anymore.